service/api_logic/auth_strategy.py

```diff
@@ -0,0 +1,96 @@
+from abc import ABC, abstractmethod
+from enum import Enum
+import bcrypt
+import requests
+from flask import current_app, request
+from oauthlib.oauth2 import WebApplicationClient
+from database.models import User
+from dto.api_input import InputUserLogInDTO
+from dto.api_output import OutputLogin
+from exept.exeptions import IncorrectUserDataError, UserDoesNotExistError, IncorrectLogInStrategyMethod, \
+    InvalidDataForGoogleLogIn
+from typing import Generic, TypeVar
+
+T = TypeVar("T")
+
+class AuthMethods(Enum):
+    SIMPLE = "simple"
+    GOOGLE = "google"
+
+
+class AuthHandler(ABC, Generic[T]):
+    def __init__(self, user_service):
+        self._user_service = user_service
+
+    @abstractmethod
+    def authenticate(self, credentials: T):
+        pass
+
+
+class AuthManager:
+    def __init__(self, user_service):
+        self._user_service = user_service
+        self.strategies = {
+            AuthMethods.SIMPLE.value: SimpleAuthHandler(user_service=self._user_service),
+            AuthMethods.GOOGLE.value: GoogleAuthHandler(user_service=self._user_service),
+        }
+
+    async def execute_log_in(self, method: str, credentials: T):
+        if method not in self.strategies:
+            raise IncorrectLogInStrategyMethod(method)
+
+        login_strategy = await self.strategies[method].authenticate(credentials)
+        return login_strategy
+
+
+class SimpleAuthHandler(AuthHandler[T]):
+    async def authenticate(self, credentials: T):
+        user = self._user_service.get_user_by_email_or_username(credentials.email_or_username, credentials.email_or_username)
+
+        if not user:
+            self._user_service._logger.warning("User does not exist")
+            raise UserDoesNotExistError(user.email)
+
+        if not bcrypt.checkpw(credentials.password_hash.encode('utf-8'), user.password_hash.encode('utf-8')):
+            self._user_service._logger.warning("Passwords do not match")
```

Comment:
That's sensitive info, let's not log it. We shouldn't specify anywhere anything about how close/far the user is from logging in. We never know who might get access to our DB or our logs. That info might help somebody hack a user in the future

Reply:
Also make sure we don't return anything like this to user (I mean through API). We don't want to become a meme, like: "This password is already used by user12345" :)